Add base class grouping and merge groupings into single operator

JonasDov opened the pull request:
Closes #245

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/ClassGrouping.ts:
```diff
@@ -0,0 +1,61 @@
+  grouped: Map<string, { class: ClassInfo; groupedNodes: Array<HierarchyNode> }>;
+}
+
+export async function createClassGroups(metadata: IMetadataProvider, nodes: HierarchyNode[]): Promise<GroupingHandlerReturn> {
```
please add `@internal` release tags to internal exported apis (I believe docs build requires that and it also makes the intention clear)

JonasDov replied:
Added

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/LabelGrouping.ts:
```diff
@@ -0,0 +1,52 @@
+    return { allNodes: nodes, groupedNodes: [] };
+  }
+  const firstNode: HierarchyNode = nodes[0].params?.grouping?.byLabel
+    ? {
+        label: nodes[0].label,
+        key: {
+          type: "label-grouping",
+          label: nodes[0].label,
+        },
+        children: [nodes[0]],
+      }
```
looks like creating a label grouping node from a single instance node could reduce code duplication

JonasDov replied:
Added a function createLabelGroupingNode()

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/BaseClassGrouping.ts:
```diff
@@ -0,0 +1,116 @@
+import { GroupingHandlerReturn } from "../Grouping";
+
+export async function getBaseClassGroupingECClasses(metadata: IMetadataProvider, nodes: HierarchyNode[]): Promise<ECClass[]> {
+  const baseEntityAndRelationshipECClassesArray = new Array<ECClass>();
```
let's just name this `classes`

JonasDov replied:
Renamed

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/BaseClassGrouping.ts:
```diff
@@ -0,0 +1,116 @@
+export async function sortByBaseClass(classes: ECClass[]) {
+  const output: ECClass[] = [];
+  const originalAmountOfClasses = classes.length;
+  while (output.length < originalAmountOfClasses) {
```
looks like this could be optimized to at least O(n^2)

JonasDov replied:
Changed

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/HierarchyProvider.ts:
```diff
@@ -201,6 +202,19 @@ export class HierarchyProvider {
 
+async function createGroupingHandlers(metadata: IMetadataProvider, nodes: HierarchyNode[]): Promise<GroupingHandlerType[]> {
+  const groupingHandlers: GroupingHandlerType[] = new Array<GroupingHandlerType>();
+  const baseClassGroupingECClasses = await getBaseClassGroupingECClasses(metadata, nodes);
+  for (const baseECClass of baseClassGroupingECClasses) {
+    groupingHandlers.push(async (allNodes: HierarchyNode[]) => {
+      return createBaseClassGroupsForSingleBaseClass(metadata, allNodes, baseECClass);
+    });
+  }
```
I think this code should be moved into `BaseClassGrouping.ts` into a function `createBaseClassGroupingHandlers(metadata, nodes)`

JonasDov replied:
Added "createBaseClassGroupingHandlers" function

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/GroupHiding.ts:
```diff
@@ -0,0 +1,115 @@
+export function applyGroupHidingParams(allNodes: HierarchyNode[]): GroupingHandlerReturn {
+  const finalGroupings: GroupingHandlerReturn = { allNodes: [], groupedNodes: [] };
+  for (const node of allNodes) {
+    if (HierarchyNode.isGroupingNode(node)) {
```
looks like we could pass it "all nodes" and "grouped nodes" separately to avoid checking if the node is a grouping node?

JonasDov replied:
applyGroupHidingParams still needs to return allNodes. So I added a different check, instead I check if the current node in allNodes array is the same as the currentNode in the groupedNodes array, if it is then it might need to be hidden.

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/GroupHiding.ts:
```diff
@@ -0,0 +1,115 @@
+ */
+function getGroupingHideOptionsFromParentNode(parentNode: HierarchyNode): [hideIfNoSiblings: boolean, hideIfOneGroupedNode: boolean] {
+  if (Array.isArray(parentNode.children)) {
+    if (HierarchyNode.isBaseClassGroupingNode(parentNode)) {
+      return getHideOptionsFromBaseClassGroupingNodes(parentNode.children);
+    }
+    if (HierarchyNode.isClassGroupingNode(parentNode)) {
+      return getHideOptionsFromClassGroupingNodes(parentNode.children);
+    }
+    if (HierarchyNode.isLabelGroupingNode(parentNode)) {
+      return getHideOptionsFromLabelGroupingNodes(parentNode.children);
+    }
```
I don't think we need 3 different functions for this. Maybe something like this would work?
```ts
function getHideOptionsFromNodeProcessingParams(nodes: HierarchyNode[], hideOptionsAccessor: (params: HierarchyNodeHandlingParams) => BaseGroupingParams | undefined): [hideIfNoSiblings: boolean, hideIfOneGroupedNode: boolean] {
  let hideIfNoSiblings = false;
  let hideIfOneGroupedNode = false;
  for (const node of nodes) {
    if (hideIfNoSiblings && hideIfOneGroupedNode) {
      break;
    }
    const params = node.params ? hideOptionsAccessor(node.params) : undefined;
    hideIfNoSiblings = !!params.hideIfNoSiblings;
    hideIfNoSiblings = !!params.hideIfOneGroupedNode;
  }
  return [hideIfNoSiblings, hideIfOneGroupedNode];
}

// usage:
if (HierarchyNode.isBaseClassGroupingNode(parentNode)) {
  return getHideOptionsFromNodeProcessingParams(parentNode.children, (p) => p.grouping?.byBaseClasses);
}
if (HierarchyNode.isClassGroupingNode(parentNode)) {
  return getHideOptionsFromNodeProcessingParams(parentNode.children, (p) => typeof p.grouping?.byClass === "object" ? p.grouping.byClass : undefined);
}
if (HierarchyNode.isLabelGroupingNode(parentNode)) {
  return getHideOptionsFromNodeProcessingParams(parentNode.children, (p) => typeof p.grouping?.byLabel === "object" ? p.grouping.byLabel : undefined);
}
```

JonasDov replied:
Changed to your suggestion, also, changed this suggestion:
```
hideIfNoSiblings = !!params.hideIfNoSiblings;
hideIfNoSiblings = !!params.hideIfOneGroupedNode;
```
to this:
```
hideIfNoSiblings ||= !!params.hideIfNoSiblings;
hideIfOneGroupedNode ||= !!params.hideIfOneGroupedNode;
```
Since only one node needs to have these options checked and not all of them.

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts:
```diff
@@ -36,11 +36,29 @@ export interface LabelGroupingNodeKey {
+ * A key for a base-class-grouping node.
+ * @beta
+ */
+export interface BaseClassGroupingNodeKey {
```
Removed

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts:
```diff
@@ -71,17 +89,52 @@ export namespace HierarchyNodeKey {
+}
+
+/** @beta */
+export interface BaseClassInfo {
```
Changed to "fullClassNames: string[]" instead of baseClassInfo: "BaseClassInfo[]"

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/HierarchyProvider.ts:
```diff
@@ -201,6 +202,19 @@ export class HierarchyProvider {
   }
 }
 
+async function createGroupingHandlers(metadata: IMetadataProvider, nodes: HierarchyNode[]): Promise<GroupingHandlerType[]> {
```
Moved

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/Common.ts:
```diff
@@ -46,8 +46,58 @@ function mergeNodeHandlingParams(
     ...(lhs?.hideInHierarchy && rhs?.hideInHierarchy ? { hideInHierarchy: true } : undefined),
-    ...(lhs?.groupByClass || rhs?.groupByClass ? { groupByClass: true } : undefined),
-    ...(lhs?.groupByLabel || rhs?.groupByLabel ? { groupByLabel: true } : undefined),
+    ...(lhs?.grouping || rhs?.grouping
```
Moved to a separate getGroupingParams() function, also added getHidingParams() function, which gets hiding parameters from specific grouping.

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/queries/NodeSelectClauseFactory.ts:
```diff
@@ -2,7 +2,6 @@
  * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
  * See LICENSE.md in the project root for license terms and full copyright notice.
  *--------------------------------------------------------------------------------------------*/
-
```
Added the empty line

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/queries/NodeSelectClauseFactory.ts:
```diff
@@ -30,10 +29,8 @@ export enum NodeSelectClauseColumnNames {
-  GroupByClass = "GroupByClass",
-  /** A flag indicating the node should be grouped by label. */
-  GroupByLabel = "GroupByLabel",
+  // /** A serialized JSON object for providing grouping information. */
```
Removed "//"

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/queries/NodeSelectClauseFactory.ts:
```diff
@@ -69,11 +66,41 @@ export interface NodeSelectClauseProps {
   mergeByLabelId?: string | ECSqlValueSelector;
 }
 
+/** @beta */
```
Removed

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/queries/NodeSelectClauseFactory.ts:
```diff
@@ -90,15 +117,47 @@ export class NodeSelectClauseFactory {
-      CAST(${createECSqlValueSelector(props.groupByLabel)} AS BOOLEAN) AS ${NodeSelectClauseColumnNames.GroupByLabel},
+      ${
+        props.grouping
+          ? `json_object(${Object.entries(props.grouping)
```
Added a separate function createGroupingSelector() which handles grouping and returns the selector string

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/Sorting.ts:
```diff
@@ -20,3 +20,10 @@ export function sortNodesByLabelOperator(nodes: Observable<HierarchyNode>): Obse
+/**
+ * @internal
+ */
+export function sortNodesByLabel(nodes: HierarchyNode[]): HierarchyNode[] {
+  return nodes.sort((lhs, rhs) => naturalCompare(lhs.label.toLocaleLowerCase(), rhs.label.toLocaleLowerCase()));
+}
```
Changed

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/Grouping.ts:
```diff
@@ -0,0 +1,86 @@
+  groupingHandlers: GroupingHandlerType[];
+}
+
+export interface GroupingHandlerReturn {
```
Renamed

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/Grouping.ts:
```diff
@@ -0,0 +1,86 @@
+  groupedNodes: HierarchyNode[];
+}
+
+export type GroupingHandlerType = (allNodes: HierarchyNode[]) => Promise<GroupingHandlerReturn>;
```
Renamed

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/Grouping.ts:
```diff
@@ -0,0 +1,86 @@
+  return groupNodes(nodes, groupingHandlers);
+}
+
+export async function groupNodes(nodes: HierarchyNode[], groupingHandlers: GroupingHandlerType[]): Promise<HierarchyNode[]> {
```
Removed export.

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/Grouping.ts:
```diff
@@ -0,0 +1,86 @@
+
+export async function groupNodes(nodes: HierarchyNode[], groupingHandlers: GroupingHandlerType[]): Promise<HierarchyNode[]> {
+  const originalNodes = nodes;
+  let allNodes = nodes;
```
Removed allNodes, using nodes instead

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/BaseClassGrouping.ts:
```diff
@@ -0,0 +1,116 @@
+  // remove grouping node if it did not have any children
+  if (Array.isArray(baseClassGroupingNode.children) && baseClassGroupingNode.children.length === 0) {
+    finalAllNodeHierarchy.splice(0, 1);
+    finalGroupedNodeHierarchy.splice(0, 1);
```
Yes, changed to push only when necessary

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/BaseClassGrouping.ts:
```diff
@@ -0,0 +1,116 @@
+          break;
+        }
+      }
+      if (classNameIsInBaseClassInfo) {
```
Added

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/BaseClassGrouping.ts:
```diff
@@ -0,0 +1,116 @@
+        const fullCurrentNodeClassName = node.key.instanceKeys[0].className;
+        const currentNodeECClass = await getClass(metadata, fullCurrentNodeClassName);
+        if (await currentNodeECClass.is(baseECClass)) {
+          if (finalAllNodeHierarchy.length > 0 && Array.isArray(baseClassGroupingNode.children)) {
```
We don't. Removed.

grigasp commented on packages/full-stack-tests/src/hierarchy-builder/grouping/BaseClassGrouping.test.ts:
```diff
@@ -0,0 +1,764 @@
+import { NodeValidators, validateHierarchy } from "../HierarchyValidation";
+
+describe("Stateless hierarchy builder", () => {
+  describe("BaseClassGrouping", () => {
```
Full stack tests test situations rather than APIs, so `describe` blocks should be in proper language, e.g. "Base class grouping" here.

JonasDov replied:
Changed.

grigasp commented on packages/full-stack-tests/src/hierarchy-builder/grouping/BaseClassGrouping.test.ts:
```diff
@@ -0,0 +1,764 @@
+      await terminate();
+    });
+
+    function createProvider(props: { imodel: IModelConnection; hierarchy: IHierarchyLevelDefinitionsFactory }) {
+      const { imodel, hierarchy } = props;
+      const schemas = new SchemaContext();
+      schemas.addLocater(new ECSchemaRpcLocater(imodel.getRpcProps()));
+      const metadataProvider = createMetadataProvider(schemas);
+      return new HierarchyProvider({
+        metadataProvider,
+        hierarchyDefinition: hierarchy,
+        queryExecutor: createECSqlQueryExecutor(imodel),
+      });
+    }
```
This is repeated across all test suites - let's move it somewhere so it can be shared.

JonasDov replied:
Moved it to `IModelUtils.ts` file

grigasp commented on packages/full-stack-tests/src/hierarchy-builder/grouping/BaseClassGrouping.test.ts:
```diff
@@ -0,0 +1,764 @@
+      });
+    }
+    describe("does not group", () => {
+      let sharedIModel: IModelConnection;
+      let sharedKeys: any;
+      beforeEach(async function () {
+        const { imodel, ...keys } = await buildIModel(this, async (builder) => {
+          const rootSubject = { className: subjectClassName, id: IModel.rootSubjectId };
+          const childSubject1 = insertSubject({ builder, codeValue: "A1", parentId: rootSubject.id });
+          const childSubject2 = insertSubject({ builder, codeValue: "A2", parentId: rootSubject.id });
+          const childPartition3 = insertPhysicalPartition({ builder, codeValue: "B3", parentId: rootSubject.id });
+          const childPartition4 = insertPhysicalPartition({ builder, codeValue: "B4", parentId: rootSubject.id });
+          const childPartition5 = insertPhysicalPartition({ builder, codeValue: "B5", parentId: rootSubject.id });
+          return { rootSubject, childSubject1, childSubject2, childPartition3, childPartition4, childPartition5 };
+        });
+        sharedKeys = keys;
+        sharedIModel = imodel;
+      });
```
I think for these tests you could use an empty iModel (which has 1 root `Subject`).

JonasDov replied:
Using an empty iModel for these tests:
- doesn't create grouping nodes if provided classes aren't base for node class
- doesn't create grouping nodes if provided classes aren't of entity or relationship type
- creates grouping nodes if provided class is base for node class

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/Common.ts:
```diff
@@ -46,12 +46,57 @@ function mergeNodeHandlingParams(
+  };
+}
+
+function getHidingParams(lhsGroupingHiding: BaseGroupingParams | undefined, rhsGroupingHiding: BaseGroupingParams | undefined): BaseGroupingParams {
```
Renamed.

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/Common.ts:
```diff
@@ -46,12 +46,57 @@ function mergeNodeHandlingParams(
   };
 }
 
+function getGroupingParams(lhsGrouping: GroupingParams | undefined, rhsGrouping: GroupingParams | undefined): GroupingParams {
```
Renamed.

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/Common.ts:
```diff
@@ -46,12 +46,57 @@ function mergeNodeHandlingParams(
     ...(lhs?.hideInHierarchy && rhs?.hideInHierarchy ? { hideInHierarchy: true } : undefined),
-    ...(lhs?.groupByClass || rhs?.groupByClass ? { groupByClass: true } : undefined),
-    ...(lhs?.groupByLabel || rhs?.groupByLabel ? { groupByLabel: true } : undefined),
+    ...(lhs?.grouping || rhs?.grouping ? getGroupingParams(lhs?.grouping, rhs?.grouping) : undefined),
```
Missed that, added { grouping: mergeNodeGroupingParams(lhs?.grouping, rhs?.grouping) }

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/Common.ts:
```diff
@@ -46,12 +46,57 @@ function mergeNodeHandlingParams(
+              : true,
+        }
+      : undefined),
+    ...(lhsGrouping?.byBaseClasses?.fullClassNames || rhsGrouping?.byBaseClasses?.fullClassNames || undefined
```
Removed.

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/BaseClassGrouping.ts:
```diff
@@ -0,0 +1,122 @@
+      const fullCurrentNodeClassName = node.key.instanceKeys[0].className;
+      const currentNodeECClass = await getClass(metadata, fullCurrentNodeClassName);
+      if (await currentNodeECClass.is(baseECClass)) {
+        if (Array.isArray(baseClassGroupingNode.children)) {
```
I used the `assert(Array.isArray(baseClassGroupingNode.children))` and removed the `isArray` check

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/Grouping.ts:
```diff
@@ -0,0 +1,100 @@
+async function groupNodesFromHandlerCreator(
+  metadata: IMetadataProvider,
+  nodes: HierarchyNode[],
+  groupingHandlerCreator: GroupingHandlerCreator,
```
Renamed.

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/Grouping.ts:
```diff
@@ -0,0 +1,100 @@
+/** @internal */
+export type GroupingHandler = (allNodes: HierarchyNode[]) => Promise<GroupingHandlerResult>;
+
+async function groupNodesFromHandlerCreator(
```
Moved it to the `createGroupingOperator`: 
`mergeMap((resolvedNodes) =>
        from(createGroupingHandlers(metadata, resolvedNodes)).pipe(mergeMap((groupingHandlers) => from(groupNodes(resolvedNodes, groupingHandlers)))),
      ),`

JonasDov commented on packages/full-stack-tests/src/hierarchy-builder/grouping/ClassGrouping.test.ts:
```diff
@@ -0,0 +1,142 @@
+            },
+          ];
+        } else if (HierarchyNode.isInstancesNode(parentNode) && parentNode.label === "root subject") {
+          return [
+            {
+              fullClassName: `BisCore.InformationContentElement`,
+              query: {
+                ecsql: `
+                  SELECT ${await selectClauseFactory.createSelectClause({
+                    ecClassId: { selector: `this.ECClassId` },
+                    ecInstanceId: { selector: `this.ECInstanceId` },
+                    nodeLabel: { selector: `this.UserLabel` },
+                    grouping: {
+                      byClass: true,
+                    },
+                  })}
+                  FROM (
+                    SELECT ECClassId, ECInstanceId, UserLabel, Parent
+                    FROM ${subjectClassName}
+                    UNION ALL
+                    SELECT ECClassId, ECInstanceId, UserLabel, Parent
+                    FROM ${physicalPartitionClassName}
+                  ) AS this
+                  WHERE this.Parent.Id = (${IModel.rootSubjectId})
+                `,
+              },
+            },
+          ];
```
Moved all the queries to the root level in grouping tests.

JonasDov commented on packages/full-stack-tests/src/hierarchy-builder/grouping/GroupHiding.test.ts:
```diff
@@ -0,0 +1,182 @@
+      });
+    }
+
+    it("groups by base class, class and label and hides some groups", async function () {
```
Added the tests you suggested, and 6 more:
- doesn't hide base class groups when there are siblings
- doesn't hide base class groups when there's more than 1 grouped node
- doesn't hide class groups when there are siblings
- doesn't hide class groups when there's more than 1 grouped node
- doesn't hide label groups when there are siblings
- doesn't hide label groups when there's more than 1 grouped node

JonasDov commented on packages/full-stack-tests/src/hierarchy-builder/grouping/BaseClassGrouping.test.ts:
```diff
@@ -0,0 +1,764 @@
+                  `,
+                  },
+                },
+              ];
```
Using only subject nodes in these tests:
- doesn't create grouping nodes if provided classes aren't base for node class
- doesn't create grouping nodes if provided classes aren't of entity or relationship type
- creates grouping nodes if provided class is base for node class
Using `Physical Partition` for these tests:
- creates multiple grouping nodes if provided base classes are base for node and for provided other base class
- creates different grouping nodes if nodes of the same class have different base classes provided
Using `Physical Partition` and `Subject` for this test:
- groups nodes of different classes if they share the same base class

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/GroupHiding.ts:
```diff
@@ -8,30 +8,25 @@ import { GroupingHandlerResult, GroupingType } from "../Grouping";
-    }
-    ++lastGroupingNodeIndex;
-    if (Array.isArray(node.children) && (props.allNodes.length === 1 || node.children.length === 1)) {
+  const finalGroupings: GroupingHandlerResult = { grouped: [], ungrouped: [], groupingType: props.groupingType };
```
```suggestion
  const finalGroupings = { ...props, grouped: [] };
```

and then we don't need the `finalGroupings.ungrouped.push(...props.ungrouped);` line at the bottom

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/LabelGrouping.ts:
```diff
@@ -0,0 +1,58 @@
+    groupedNodes: isLabelGrouping ? [firstNode] : [],
+    ungroupedNodes: isLabelGrouping ? [] : [firstNode],
+    groupingType: "label",
+  };
```
Applied this suggestion. Changed manually, because your comment was a little bit outdated.

JonasDov commented on packages/full-stack-tests/src/IModelUtils.ts:
```diff
@@ -385,3 +389,15 @@ export function insertExternalSourceAspect(
   return { className, id };
 }
+
+export function createProvider(props: { imodel: IModelConnection; hierarchy: IHierarchyLevelDefinitionsFactory }) {
```
Moved to hierarchy-builder/Utils.ts

JonasDov commented on packages/full-stack-tests/src/hierarchy-builder/grouping/BaseClassGrouping.test.ts:
```diff
@@ -281,98 +140,53 @@ describe("Stateless hierarchy builder", () => {
+        const childPartition1 = insertPhysicalPartition({ builder, codeValue: "B1", parentId: IModel.rootSubjectId, userLabel: "test" });
+        const childPartition2 = insertPhysicalPartition({ builder, codeValue: "B2", parentId: IModel.rootSubjectId, userLabel: "test" });
+        const childPartition3 = insertPhysicalPartition({ builder, codeValue: "B3", parentId: IModel.rootSubjectId, userLabel: "test" });
+        return { childPartition1, childPartition2, childPartition3 };
```
Changed

JonasDov commented on packages/full-stack-tests/src/hierarchy-builder/grouping/BaseClassGrouping.test.ts:
```diff
@@ -388,113 +202,77 @@ describe("Stateless hierarchy builder", () => {
+        const childPartition2 = insertPhysicalPartition({ builder, codeValue: "B2", parentId: IModel.rootSubjectId, userLabel: "test" });
+        const childPartition3 = insertPhysicalPartition({ builder, codeValue: "B3", parentId: IModel.rootSubjectId, userLabel: "test" });
+        const childPartition4 = insertPhysicalPartition({ builder, codeValue: "B4", parentId: IModel.rootSubjectId, userLabel: "test" });
+        return { childPartition1, childPartition2, childPartition3, childPartition4 };
```
Changed

JonasDov commented on packages/full-stack-tests/src/hierarchy-builder/grouping/BaseClassGrouping.test.ts:
```diff
@@ -537,123 +312,91 @@ describe("Stateless hierarchy builder", () => {
+        const childPartition4 = insertPhysicalPartition({ builder, codeValue: "B4", parentId: IModel.rootSubjectId, userLabel: "test" });
+        const childPartition5 = insertPhysicalPartition({ builder, codeValue: "B5", parentId: IModel.rootSubjectId, userLabel: "test" });
+        return { childSubject1, childSubject2, childPartition3, childPartition4, childPartition5 };
+      });
```
Changed

JonasDov commented on packages/full-stack-tests/src/hierarchy-builder/grouping/GroupHiding.test.ts:
```diff
@@ -31,52 +28,46 @@ describe("Stateless hierarchy builder", () => {
+                      grouping: {
+                        byBaseClasses: {
+                          fullClassNames: ["BisCore.InformationReferenceElement"],
+                          hideIfNoSiblings: true,
```
Created `createHierarchyWithSpecifiedGrouping` function which is used for all tests in this file

JonasDov commented on packages/full-stack-tests/src/hierarchy-builder/grouping/GroupHiding.test.ts:
```diff
@@ -87,14 +78,172 @@ describe("Stateless hierarchy builder", () => {
+      it("doesn't hide base class groups when there's more than 1 grouped node", async function () {
+        const { imodel, ...keys } = await buildIModel(this, async (builder) => {
+          const childSubject1 = insertSubject({ builder, codeValue: "A1", parentId: IModel.rootSubjectId, userLabel: groupName });
+          const childSubject2 = insertSubject({ builder, codeValue: "A2", parentId: IModel.rootSubjectId, userLabel: groupName });
```
Not really, removed.

grigasp commented on packages/hierarchy-builder/src/hierarchy-builder/queries/NodeSelectClauseFactory.ts:
```diff
@@ -71,13 +71,12 @@ export interface NodeSelectClauseProps {
-type ECSqlSelectClauseGroupingParams = LabelGroupingParams & ClassGroupingParams;
-
-interface LabelGroupingParams {
+/**
+ * A data structure for defining nodes' grouping requirements.
+ * @beta
+ */
+export interface ECSqlSelectClauseGroupingParams {
```
```suggestion
interface ECSqlSelectClauseGroupingParams {
```

grigasp replied:
nvm, I see it's used by `NodeSelectClauseProps`

grigasp commented on packages/hierarchy-builder/src/test/Utils.ts:
```diff
@@ -96,3 +100,7 @@ export function createGetClassStub(schemas: IMetadataProvider) {
   return { getClass: stub, stubClass };
 }
+
+export async function isMock(className: string): Promise<boolean> {
+  return className.includes("Parent");
+}
```
I don't think this should be used at all:
- The name is way too obscure
- The behavior kind of depends on how the classes are set up in each test

JonasDov replied:
Removed

grigasp commented on packages/hierarchy-builder/src/test/internal/operators/grouping/BaseClassGrouping.test.ts:
```diff
@@ -0,0 +1,253 @@
+    stubClass({
+      schemaName: "TestSchema",
+      className: "TestParentClass",
+      classLabel: "TestSchema.TestParentClass",
```
The label property should be in label format rather than just set to full class name. However, I don't think you need it at all in this test.

JonasDov replied:
Removed

grigasp reviewed (commented):
I just now looked at the unit tests. They're testing good situations, but seem to be overly complex - to test one grouping handler we need to pass the input through the operator + all the handlers. My suggestion:
- Change `createGroupingOperator` function to optionally take an array of grouping handlers. Default to whatever is used now. This will allow us to test the operator behavior with stub handlers.
- Export each grouping handler as `@internal` API. This will allow us to test each of them separately.

JonasDov commented:
> I just now looked at the unit tests. They're testing good situations, but seem to be overly complex - to test one grouping handler we need to pass the input through the operator + all the handlers. My suggestion:
> 
> * Change `createGroupingOperator` function to optionally take an array of grouping handlers. Default to whatever is used now. This will allow us to test the operator behavior with stub handlers.
> * Export each grouping handler as `@internal` API. This will allow us to test each of them separately.

- Added suggested changes to `createGroupingOperator`.
- Changed unit tests to test grouping functions instead of the `createGroupingOperator`

grigasp commented on packages/hierarchy-builder/src/test/internal/operators/Grouping.test.ts:
```diff
@@ -0,0 +1,207 @@
+        createTestNode({
+          key: { type: "instances", instanceKeys: [{ className: "TestSchema.A", id: "0x1" }] },
+          label: "1",
+          params: {
+            grouping: {
+              byBaseClasses: {
+                fullClassNames: ["TestSchema.TestParentClassA", "TestSchema.TestParentClassAA"],
+              },
+              byClass: true,
+              byLabel: true,
+            },
+          },
```
I don't think we need to pass grouping params - the grouping operator is using stubbed handlers which don't care about them. Applies to more tests below.

JonasDov replied:
Removed

grigasp commented on packages/hierarchy-builder/src/test/internal/operators/Grouping.test.ts:
```diff
@@ -0,0 +1,207 @@
+      const result = await getObservableResult(
+        from(nodes).pipe(
+          createGroupingOperator(metadata, [
+            async (allNodes) => {
+              return { grouped: [], ungrouped: allNodes, groupingType: "label" } as GroupingHandlerResult;
+            },
+            async (allNodes) => {
+              return { grouped: [], ungrouped: allNodes, groupingType: "class" } as GroupingHandlerResult;
+            },
```
Could this be simplified to the following?
```suggestion
            async (allNodes) => ({ grouped: [], ungrouped: allNodes, groupingType: "label" }),
            async (allNodes) => ({ grouped: [], ungrouped: allNodes, groupingType: "class" }),
```

grigasp commented on packages/hierarchy-builder/src/test/internal/operators/Grouping.test.ts:
```diff
@@ -0,0 +1,207 @@
+          ]),
+        ),
+      );
+      expect(applyGroupingHidingParamsStub.callCount).to.eq(3);
```
should we also check what arguments were passed to each call of this stub and also the grouping handlers?

JonasDov replied:
Added checks.

grigasp commented on packages/hierarchy-builder/src/test/internal/operators/Grouping.test.ts:
```diff
@@ -0,0 +1,207 @@
+    let createLabelGroupsStub: sinon.SinonStub;
+    before(() => {
+      createBaseClassGroupingHandlersStub = sinon.stub(baseClassGrouping, "createBaseClassGroupingHandlers").resolves([] as GroupingHandler[]);
+      createClassGroupsStub = sinon.stub(classGrouping, "createClassGroups").resolves({ grouped: [], ungrouped: [], groupingType: "class" });
+      createLabelGroupsStub = sinon.stub(labelGrouping, "createLabelGroups").resolves({ grouped: [], ungrouped: [], groupingType: "label" });
```
doesn't look like you need the stubs to resolve anything

JonasDov replied:
Removed.

grigasp commented on packages/hierarchy-builder/src/test/internal/operators/Grouping.test.ts:
```diff
@@ -0,0 +1,207 @@
+      const result = await createGroupingHandlers(metadata, []);
+      expect(createBaseClassGroupingHandlersStub.callCount).to.eq(1);
+      expect(result.length).to.eq(2);
+      expect(createClassGroupsStub.callCount).to.eq(0);
+      await result[0]([]);
+      expect(createClassGroupsStub.callCount).to.eq(1);
```
would an `expect(result[0]).to.eq(createClassGroupsStub)` check work here?

JonasDov replied:
No it would not, result[i] returns a function that calls the stub, not the stub itself:
![image](https://github.com/iTwin/presentation/assets/100586436/03035d91-5182-46ce-b217-c32e78aba93b)

grigasp commented on packages/hierarchy-builder/src/test/internal/operators/Grouping.test.ts:
```diff
@@ -0,0 +1,207 @@
+
+    it("creates grouping handlers in class -> label grouping order", async () => {
+      const result = await createGroupingHandlers(metadata, []);
+      expect(createBaseClassGroupingHandlersStub.callCount).to.eq(1);
```
I think it's important to also check that `createBaseClassGroupingHandlersStub` is called with the nodes we pass to `createGroupingHandlers`

JonasDov replied:
Added.

grigasp commented on packages/hierarchy-builder/src/test/internal/operators/grouping/BaseClassGrouping.test.ts:
```diff
@@ -0,0 +1,298 @@
+      });
+
+      const result = await baseClassGrouping.getBaseClassGroupingECClasses(metadataProvider, nodes);
+      expect(result[0].fullName).to.eq("TestSchema.TestParentClass1");
+      expect(result[1].fullName).to.eq("TestSchema.TestParentClass3");
+      expect(result[2].fullName).to.eq("TestSchema.TestParentClass2");
```
Could we:
- simplify class names to "Class1", "Class2", "Class3"
- make "1" the most base class and "3" the most derived one

JonasDov replied:
Changed to your suggestion.

grigasp commented on packages/hierarchy-builder/src/test/internal/operators/grouping/BaseClassGrouping.test.ts:
```diff
@@ -0,0 +1,298 @@
+
+describe("BaseClassGrouping", () => {
+  const metadataProvider = {} as unknown as IMetadataProvider;
+  const ecClass = {
+    fullName: "TestSchema.TestParentClass",
+    label: "TestParentClass",
+  } as unknown as ECClass;
```
I think it's better to define this in the tests that use it - otherwise, when reading a test, we have to scroll up to see what class is this

JonasDov replied:
Changed.

grigasp commented on packages/hierarchy-builder/src/test/internal/operators/grouping/LabelGrouping.test.ts:
```diff
@@ -0,0 +1,107 @@
+      }),
+    ];
+    const result = await createLabelGroups(nodes);
+    expect(result.ungrouped).to.deep.eq([nodes[0]]);
```
```suggestion
    expect(result.ungrouped).to.deep.eq(nodes);
```

JonasDov commented on packages/hierarchy-builder/src/hierarchy-builder/internal/operators/Grouping.ts:
```diff
@@ -19,13 +19,19 @@ const OPERATOR_NAME = "Grouping";
+        if (groupingHandlers === undefined) {
+          return from(createGroupingHandlers(metadata, resolvedNodes)).pipe(
+            mergeMap((createdGroupingHandlers) => from(groupNodes(resolvedNodes, createdGroupingHandlers))),
+          );
```
Changed to:
![image](https://github.com/iTwin/presentation/assets/100586436/e08935f1-b82b-461b-95dd-9c55486a2542)
`createGroupingHandlers` is an async function, so this suggestion would not work.
